File: server/controllers/user-controller.js
const asyncHandler = require("express-async-handler");
const User = require("../models/userModel");
const bcrypt = require("bcryptjs");
const generateToken = require('../config/genToken')

// @desc    Register a new user
// @route   POST /api/user
const registerUser = asyncHandler(async (req, res) => {
  const { name, email, password } = req.body;

  if (!name || !email || !password) throw new Error("Please fill all fields");

  const userExists = await User.findOne({ email });
  if (userExists) throw new Error("User already exists");

  const salt = await bcrypt.genSalt(10);
  const hashedPassword = await bcrypt.hash(password, salt);

  const user = await User.create({
    name,
    email,
    password: hashedPassword,
  });

  res.status(201).json({
    _id: user._id,
    name: user.name,
    email: user.email,
    token: generateToken(user._id),
  });
});

// @desc    Authenticate user & get token
// @route   POST /api/user/login
const loginUser = asyncHandler(async (req, res) => {
  const { email, password } = req.body;

  const user = await User.findOne({ email });
  if (!user) throw new Error("User not found");

  if (user.isOAuth) throw new Error("Please log in using Google");

  const isMatch = await bcrypt.compare(password, user.password);
  if (!isMatch) throw new Error("Invalid credentials");

  res.json({
    _id: user._id,
    name: user.name,
    email: user.email,
    token: generateToken(user._id),
  });
});

// @desc    Google OAuth Signup
// @route   POST /api/user/google-signup
const googleSignup = asyncHandler(async (req, res) => {
  const { email, username, googleId } = req.body;

  if (!email || !username || !googleId) throw new Error("Missing required fields");

  const existingUser = await User.findOne({ email });
  if (existingUser) throw new Error("User already exists. Please log in.");

  const user = await User.create({
    name: username,
    email,
    password: googleId, // not used, but required by schema
    isOAuth: true,
  });

  res.status(201).json({
    _id: user._id,
    name: user.name,
    email: user.email,
    token: generateToken(user._id),
  });
});

// @desc    Google OAuth Login
// @route   POST /api/user/google-login
const googleLogin = asyncHandler(async (req, res) => {
  const { email } = req.body;

  if (!email) throw new Error("Missing email");

  const user = await User.findOne({ email });

  if (!user || !user.isOAuth) throw new Error("Account not found. Please sign up first.");

  res.json({
    _id: user._id,
    name: user.name,
    email: user.email,
    token: generateToken(user._id),
  });
});

// @desc    Get all users (for testing or search)
// @route   GET /api/user
const allUsers = asyncHandler(async (req, res) => {
  const keyword = req.query.search
    ? {
        name: { $regex: req.query.search, $options: "i" },
      }
    : {};

  const users = await User.find(keyword).find({ _id: { $ne: req.user._id } });
  res.json(users);
});

// @desc    Get user by ID
// @route   GET /api/user/:id
const getUserById = asyncHandler(async (req, res) => {
  const user = await User.findById(req.params.id).select("-password");

  if (!user) {
    res.status(404);
    throw new Error("User not found");
  }

  res.json(user);
});

module.exports = { registerUser, loginUser, googleSignup, googleLogin, allUsers, getUserById};

File: server/app/controllers/user_controller.py
from fastapi import HTTPException
from datetime import datetime, timezone

from bson import ObjectId

from app.schemas.user import UserCreate, UserLogin
from app.schemas.auth import AuthResponse, AuthResult, GoogleSignupRequest, GoogleLoginRequest

from app.core.google_auth import verify_google_token

from app.core.database import user_collection
from app.core.security import (
    hash_password,
    create_access_token,
    verify_password
)

async def register_user(payload: UserCreate):
    existing_user = await user_collection.find_one({
            "email": payload.email
        })
    
    if existing_user:
        raise HTTPException(
            status_code=400,
            detail="User already exists"
        )
        
    hashed_password = hash_password(payload.password)
    
    user_data = {
        "name":payload.name,
        "email":payload.email,
        "password":hashed_password,
        "isOAuth": False,
        "createdAt": datetime.now(timezone.utc),
        "updatedAt": datetime.now(timezone.utc)
    }
    
    result = await user_collection.insert_one(user_data)
    
    token = create_access_token(str(result.inserted_id))
    
    return AuthResult(
        user=AuthResponse(
            id=str(result.inserted_id),
            name=payload.name,
            email=payload.email
        ),
        token=token
    )
    
async def login_user(payload: UserLogin):
    user = await user_collection.find_one({
            "email": payload.email,
            "isOAuth": False
        })
    
    if not user:
        raise HTTPException(
            status_code=401,
            detail="Invalid credentials"
        )
            
    if not verify_password(payload.password, user["password"]):
        raise HTTPException(
            status_code=401,
            detail="Invalid credentials"
        )
        
    token = create_access_token(str(user["_id"]))
    
    return AuthResult(
        user=AuthResponse(
            id=str(user["_id"]),
            name=user["name"],
            email=user["email"]
        ),
        token=token
    )
    
async def google_signup(payload: GoogleSignupRequest):
    google_user = await verify_google_token(payload.access_token)
    
    existing_user = await user_collection.find_one({
        "email": google_user["email"]
    })
    
    if existing_user:
        raise HTTPException(
            status_code=400,
            detail="User already exists"
        )
        
    user_data = {
        "name": google_user["name"],
        "email": google_user["email"],
        "googleId": google_user["googleId"],
        "isOAuth": True,
        "createdAt": datetime.now(timezone.utc),
        "updatedAt": datetime.now(timezone.utc)
    }
    
    result = await user_collection.insert_one(user_data)
    
    token = create_access_token(str(result.inserted_id))
    
    return AuthResult(
        user=AuthResponse(
            id=str(result.inserted_id),
            name=google_user["name"],
            email=google_user["email"],
            picture=google_user["picture"]
        ),
        token=token
    )
    
async def google_login(payload: GoogleLoginRequest):
    google_user = await verify_google_token(payload.access_token)
    
    user = await user_collection.find_one({
        "email": google_user["email"],
        "googleId": google_user["googleId"],
        "isOAuth": True
    })
    
    if not user:
        raise HTTPException(
            status_code=401,
            detail="Invalid credentials"
        )
        
    token = create_access_token(str(user["_id"]))
    
    return AuthResult(
        user=AuthResponse(
            id=str(user["_id"]),
            name=user["name"],
            email=user["email"],
            picture=google_user["picture"]
        ),
        token=token
    )
    
async def search_users(search: str, current_user):
    query = {}
    
    if search:
        query["name"] = {
            "$regex": search,
            "$options": "i",
        }
        
    users = await user_collection.find(query).limit(10).to_list(10)
    
    users = [
        user 
        for user in users
        if str(user["_id"]) != str(current_user["_id"])
    ]
    
    for user in users:
        user["_id"] = str(user["_id"])
        user.pop("password", None)
        user.pop("googleId", None)
        
    return users

async def get_user_by_id(user_id: str):
    user = await user_collection.find_one({
        "_id": ObjectId(user_id)
    })
    
    if not user:
        raise HTTPException(
            status_code=400,
            detail="User not found"
        )
    
    user["_id"] = str(user["_id"])
    user.pop("password", None)  
    
    return user
